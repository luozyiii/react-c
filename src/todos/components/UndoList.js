import React, { Component } from 'react';

export default class UndoList extends Component {
  constructor(props) {
    super(props);
    this.state = {};
  }

  render() {
    const { list, deleteItem, changeStatus, handleBlur, changeValue } = this.props;
    return (
      <div className="undo-list">
        <div className="undo-list-title">
          正在进行
          <div className="undo-list-count" data-test="count">
            {list.length}
          </div>
        </div>
        <ul className="undo-list-content">
          {list.map((item, index) => (
            <li
              className="undo-list-item"
              data-test="list-item"
              key={index}
              onClick={() => {
                changeStatus(index);
              }}
            >
              {item.status === 'div' ? item.value : <input className="undo-list-input" value={item.value} data-test="input" onBlur={() => handleBlur(index)} onChange={(e) => changeValue(index, e.target.value)} />}
              <div
                className="undo-list-delete"
                data-test="delete-item"
                onClick={(e) => {
                  e.stopPropagation();
                  deleteItem(index);
                }}
              >
                -
              </div>
            </li>
          ))}
        </ul>
      </div>
    );
  }
}
