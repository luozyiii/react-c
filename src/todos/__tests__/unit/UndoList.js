import React from 'react';
import { shallow } from 'enzyme';
import { expect } from '@jest/globals';

import { findTestWrapper } from '../../../utils/testUtils';

import UndoList from '../../components/UndoList';

describe('UndoList 组件', () => {
  it('渲染样式正常', () => {
    const wrapper = shallow(<UndoList list={[]} />);
    expect(wrapper).toMatchSnapshot();
  });

  it('列表数据为空数组时， count 数目为0，列表无内容', () => {
    const wrapper = shallow(<UndoList list={[]} />);
    const countElem = findTestWrapper(wrapper, 'count');
    const listItem = findTestWrapper(wrapper, 'list-item');
    expect(countElem.text()).toEqual('0');
    expect(listItem.length).toEqual(0);
  });

  it('列表数据有内容时， count 数目显示数据长度，列表不为空', () => {
    const listData = [
      { status: 'div', value: '学习Jest' },
      { status: 'div', value: '学习React组件' },
      { status: 'div', value: '学习TDD' },
    ];
    const wrapper = shallow(<UndoList list={listData} />);
    const countElem = findTestWrapper(wrapper, 'count');
    const listItem = findTestWrapper(wrapper, 'list-item');
    expect(countElem.text()).toEqual('3');
    expect(listItem.length).toEqual(3);
  });

  it('列表数据有内容时， 存在删除按钮', () => {
    const listData = [
      { status: 'div', value: '学习Jest' },
      { status: 'div', value: '学习React组件' },
      { status: 'div', value: '学习TDD' },
    ];
    const wrapper = shallow(<UndoList list={listData} />);
    const deleteItem = findTestWrapper(wrapper, 'delete-item');
    expect(deleteItem.length).toEqual(3);
  });

  it('列表数据有内容时， 点击某个删除按钮，会调用删除方法', () => {
    const listData = [
      { status: 'div', value: '学习Jest' },
      { status: 'div', value: '学习React组件' },
      { status: 'div', value: '学习TDD' },
    ];
    const fn = jest.fn();
    const index = 1;
    const wrapper = shallow(<UndoList deleteItem={fn} list={listData} />);
    const deleteItem = findTestWrapper(wrapper, 'delete-item');
    deleteItem.at(index).simulate('click');
    expect(fn).toHaveBeenLastCalledWith(index);
  });

  it('当某一项被点击时，触发 执行 changeStatus 函数', () => {
    const listData = [
      { status: 'div', value: '学习Jest' },
      { status: 'div', value: '学习React组件' },
      { status: 'div', value: '学习TDD' },
    ];
    const fn = jest.fn();
    const index = 1;
    const wrapper = shallow(<UndoList changeStatus={fn} list={listData} />);
    const listItem = findTestWrapper(wrapper, 'list-item');
    listItem.at(index).simulate('click');
    expect(fn).toHaveBeenLastCalledWith(index);
  });

  it('当某一项状态是 input 时，展示输入框', () => {
    const listData = [
      { status: 'input', value: '学习Jest' },
      { status: 'div', value: '学习React组件' },
      { status: 'div', value: '学习TDD' },
    ];
    const wrapper = shallow(<UndoList list={listData} />);
    const inputItem = findTestWrapper(wrapper, 'input');
    expect(inputItem.length).toBe(1);
  });

  it('当某一输入框失去焦点时，触发执行 handleBlur 函数', () => {
    const listData = [
      { status: 'input', value: '学习Jest' },
      { status: 'div', value: '学习React组件' },
      { status: 'div', value: '学习TDD' },
    ];
    const fn = jest.fn();
    const wrapper = shallow(<UndoList handleBlur={fn} list={listData} />);
    const inputItem = findTestWrapper(wrapper, 'input');
    inputItem.simulate('blur');
    expect(fn).toHaveBeenLastCalledWith(0);
  });

  it('当某一输入框变更时，触发执行 changeValue 函数', () => {
    const listData = [
      { status: 'input', value: '学习Jest' },
      { status: 'div', value: '学习React组件' },
      { status: 'div', value: '学习TDD' },
    ];
    const value = '学习BDD';
    const fn = jest.fn();
    const wrapper = shallow(<UndoList changeValue={fn} list={listData} />);
    const inputItem = findTestWrapper(wrapper, 'input');
    inputItem.simulate('change', {
      target: { value },
    });
    expect(fn).toHaveBeenLastCalledWith(0, value);
  });
});
