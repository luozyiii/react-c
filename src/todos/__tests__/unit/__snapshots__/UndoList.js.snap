// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`UndoList 组件 渲染样式正常 1`] = `
<div
  className="undo-list"
>
  <div
    className="undo-list-title"
  >
    正在进行
    <div
      className="undo-list-count"
      data-test="count"
    >
      0
    </div>
  </div>
  <ul
    className="undo-list-content"
  />
</div>
`;
